nixos/modules/services/misc/paperless-ng.nix

```diff
@@ -0,0 +1,197 @@
+        message = "Paperless-ng replaces Paperless, either disable paperless or assign a new dataDir and port to one of them";
+      }
+    ];
+
+    systemd.tmpfiles.rules = [
+      "d '${cfg.dataDir}' - ${cfg.user} ${config.users.users.${cfg.user}.group} - -"
+      "d '${cfg.mediaDir}' - ${cfg.user} ${config.users.users.${cfg.user}.group} - -"
+      (if cfg.consumptionDirIsPublic then
+        "d '${cfg.consumptionDir}' 777 - - - -"
+      else
+        "d '${cfg.consumptionDir}' - ${cfg.user} ${config.users.users.${cfg.user}.group} - -"
+      )
+    ];
+
+    systemd.services.paperless-ng-consumer = {
+      description = "Paperless document consumer";
+      serviceConfig = {
+        User = cfg.user;
+        ExecStart = "${cfg.package}/bin/paperless-ng document_consumer";
+        Restart = "on-failure";
+      };
+      environment = env;
+      after = [ "systemd-tmpfiles-setup.service" ];
+      wantedBy = [ "multi-user.target" ];
+      preStart = ''
+        if [[ $(readlink ${cfg.dataDir}/paperless-ng-manage) != ${manage} ]]; then
+          ln -sf ${manage} ${cfg.dataDir}/paperless-ng-manage
+        fi
+
+        ${setupEnv}
+        # Auto-migrate on first run or if the package has changed
+        versionFile="${cfg.dataDir}/src-version"
+        if [[ $(cat "$versionFile" 2>/dev/null) != ${cfg.package} ]]; then
+          ${cfg.package}/bin/paperless-ng migrate
+          echo ${cfg.package} > "$versionFile"
+        fi
+      '';
+    };
+
+    systemd.services.paperless-ng-server = {
+      description = "Paperless document server";
+      serviceConfig = {
+        User = cfg.user;
+        ExecStart = "${cfg.package}/bin/paperless-ng qcluster";
+        Restart = "on-failure";
+      };
+      environment = env;
+      # Bind to `paperless-ng-consumer` so that the server never runs
+      # during migrations
+      bindsTo = [ "paperless-ng-consumer.service" ];
+      after = [ "paperless-ng-consumer.service" ];
+      wantedBy = [ "multi-user.target" ];
+    };
+
+    systemd.services.paperless-ng-web = {
+      description = "Paperless web server";
+      serviceConfig = {
+        User = cfg.user;
+        ExecStart = "${pkgs.python3Packages.gunicorn}/bin/gunicorn -c ${cfg.package}/lib/paperless-ng/gunicorn.conf.py paperless.asgi:application";
+        Restart = "on-failure";
+        Environment = [
+          # Set PATH and PYTHONPATH
+          "PATH=${cfg.package.path}"
+          "PYTHONPATH=${cfg.package.pythonPath}:${cfg.package}/lib/paperless-ng/src"
+        ];
+      };
+      environment = env;
+      # Bind to `paperless-ng-consumer` so that the server never runs
+      # during migrations
+      bindsTo = [ "paperless-ng-consumer.service" ];
+      after = [ "paperless-ng-consumer.service" ];
+      wantedBy = [ "multi-user.target" ];
+    };
+
+    users = optionalAttrs (cfg.user == defaultUser) {
```

Comment:
```suggestion
    users = mkIf (cfg.user == defaultUser) {
```

Reply:
What's the rationale? `mkIf` is slower but its payload doesn't conflict with evaluating the condition, so it's not needed to avoid infinite recursion.

Reply:
I didn't know it's slower. I guess it doesn't matter too much.

Reply:
It's (in this case, insignificantly) slower because the payload [gets transformed](https://github.com/NixOS/nixpkgs/blob/73e0dd4b291d961dd741c53c95b6c497b1c533d4/lib/modules.nix#L589) to 
```
 {
      users.${defaultUser} = {
        group = mkIf condition defaultUser;
        uid = mkIf condition config.ids.uids.paperless;
        home = mkIf condition cfg.dataDir;
      };
      ...
}
```
I'd just avoid `mkIf` unless explicitly needed.